src/ai/backend/storage/client/s3.py

```diff
@@ -1,14 +1,68 @@
+from __future__ import annotations
+
 import logging
-from collections.abc import AsyncIterable
-from typing import Any, AsyncIterator, Iterable, Optional
+from collections.abc import AsyncIterator
+from typing import Any, Iterable, Optional, override
 
 import aioboto3
 
 from ai.backend.common.dto.storage.response import ObjectMetaResponse, PresignedUploadObjectResponse
+from ai.backend.common.types import StreamReader
 
 logger = logging.getLogger(__name__)
 
 
+class S3DownloadStreamReader(StreamReader):
+    def __init__(
+        self,
+        s3_key: str,
+        bucket_name: str,
+        chunk_size: int,
+        endpoint_url: str,
+        region_name: Optional[str],
+        aws_access_key_id: Optional[str],
+        aws_secret_access_key: Optional[str],
+        content_type: Optional[str],
```

Comment:
How about managing related values together, such as a dataclass for the s3 target and a dataclass for the secret?